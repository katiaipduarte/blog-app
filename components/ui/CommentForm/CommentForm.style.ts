import styled from 'styled-components';

export const CommentFormContainer = styled.form`
  margin: 2rem auto;

  p {
    font-size: 1.25rem;
    line-height: 1.75rem;
    color: ${({ theme }) => theme.grey};
  }

  div {
    margin: 2rem 0;

    input[type='text'],
    textarea {
      width: 100% !important;
      border-radius: 0.4rem;
      border-color: ${({ theme }) => theme.grey};
      box-sizing: border-box;
      margin: 1rem 0;
      box-shadow: 0 0 20px 0 rgba(0, 0, 0, 0.2), 0 5px 5px 0 rgba(0, 0, 0, 0.24);
      padding: 0.5rem;

      &:focus-visible,
      &:focus:not(:focus-visible) {
        outline: none;
        border: 2px solid ${({ theme }) => theme.secondaryColor};
      }
      display: block;
    }

    input[type='text'] {
      height: 2rem;
    }

    textarea {
      height: 10rem !important;
      resize: none;
    }

    .invalid {
      border: 2px solid ${({ theme }) => theme.primaryColor} !important;
    }

    .message {
      color: ${({ theme }) => theme.primaryColor};
      font-size: 0.75rem;
      display: none;
    }

    .show {
      display: block !important;
    }
  }

  @media all and (min-width: 960px) {
    text-align: left;

    div {
      input[type='text'],
      textarea {
        width: 70% !important;
      }
    }
  }
`;
