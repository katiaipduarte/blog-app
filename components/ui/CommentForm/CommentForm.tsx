import React, { ChangeEvent, FormEvent, memo, useState } from 'react';
import { CommentFormContainer } from './CommentForm.style';

type Props = {
  postId: number;
  parentCommentId: number | null;
  user: string;
  commentAdded: () => Promise<void>;
};

const CommentForm = (props: Props): JSX.Element => {
  const { postId, parentCommentId, user, commentAdded } = props;
  const [formData, setFormData] = useState({
    content: '',
    user: ''
  });
  const [invalidUser, setInvalidUser] = useState<boolean>(false);
  const [invalidComment, setInvalidComment] = useState<boolean>(false);

  const handleChange = (
    event: ChangeEvent<HTMLInputElement & HTMLTextAreaElement>
  ): void => {
    const target = event.target;
    const id = target.id;
    const value = target.value;

    if (id === 'content') {
      setInvalidComment(false);
    } else {
      setInvalidUser(false);
    }

    setFormData({ ...formData, [id]: value });
  };

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (formData.content === '' || formData.user === '') {
      setInvalidComment(formData.content === '');
      setInvalidUser(formData.content === '');
      return;
    }

    await fetch(
      `${process.env.NEXT_PUBLIC_API_ENDPOINT}/posts/${postId}/comments`,
      {
        method: 'POST',
        mode: 'cors',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          ...formData,
          date: new Date().toISOString().split('T')[0],
          parent_id: parentCommentId,
          postId: postId
        })
      }
    )
      .then(() => {
        setFormData({
          content: '',
          user: ''
        });

        commentAdded();
      })
      .catch(() => {
        console.error('something went wrong while adding a comment');
      });
  };

  return (
    <CommentFormContainer onSubmit={handleSubmit}>
      <h2>Submit a comment</h2>
      {user && <p>{`Reply to ${user}`}</p>}
      <div>
        <input
          id="user"
          type="text"
          placeholder="Username"
          value={formData.user}
          onChange={handleChange}
          className={invalidUser ? 'invalid' : ''}
        />
        <span className={invalidUser ? 'show message' : 'message'}>
          Username cannot be empty
        </span>
      </div>
      <div>
        <textarea
          name="comment"
          id="content"
          value={formData.content}
          onChange={handleChange}
          placeholder="Write your comment here..."
          className={invalidComment ? 'invalid' : ''}
        />
        <span className={invalidComment ? 'show message' : 'message'}>
          Comment cannot be empty
        </span>
      </div>
      <button type="submit" className="full">
        Submit
      </button>
    </CommentFormContainer>
  );
};

export default memo(CommentForm);
