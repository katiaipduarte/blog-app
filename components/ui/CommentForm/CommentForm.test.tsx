import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import CommentForm from './CommentForm';

describe('the <CommentForm /> component', () => {
  it('should render comment form to add a comment to post', () => {
    const tree = renderer
      .create(
        <CommentForm
          postId={10}
          parentCommentId={null}
          user={''}
          commentAdded={jest.fn().mockName('mockedFunction')}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render comment form to reply to a comment', () => {
    const tree = renderer
      .create(
        <CommentForm
          postId={10}
          parentCommentId={2}
          user={'test'}
          commentAdded={jest.fn().mockName('mockedFunction')}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
