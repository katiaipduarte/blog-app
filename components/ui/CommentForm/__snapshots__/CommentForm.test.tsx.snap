// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <CommentForm /> component should render comment form to add a comment to post 1`] = `
.c0 {
  margin: 2rem auto;
}

.c0 p {
  font-size: 1.25rem;
  line-height: 1.75rem;
}

.c0 div {
  margin: 2rem 0;
}

.c0 div input[type='text'],
.c0 div textarea {
  width: 100% !important;
  border-radius: 0.4rem;
  box-sizing: border-box;
  margin: 1rem 0;
  box-shadow: 0 0 20px 0 rgba(0,0,0,0.2),0 5px 5px 0 rgba(0,0,0,0.24);
  padding: 0.5rem;
  display: block;
}

.c0 div input[type='text']:focus-visible,
.c0 div textarea:focus-visible,
.c0 div input[type='text']:focus:not(:focus-visible),
.c0 div textarea:focus:not(:focus-visible) {
  outline: none;
  border: 2px solid;
}

.c0 div input[type='text'] {
  height: 2rem;
}

.c0 div textarea {
  height: 10rem !important;
  resize: none;
}

.c0 div .invalid {
  border: 2px solid !important;
}

.c0 div .message {
  font-size: 0.75rem;
  display: none;
}

.c0 div .show {
  display: block !important;
}

@media all and (min-width:960px) {
  .c0 {
    text-align: left;
  }

  .c0 div input[type='text'],
  .c0 div textarea {
    width: 70% !important;
  }
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <h2>
    Submit a comment
  </h2>
  
  <div>
    <input
      className=""
      id="user"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <span
      className="message"
    >
      Username cannot be empty
    </span>
  </div>
  <div>
    <textarea
      className=""
      id="content"
      name="comment"
      onChange={[Function]}
      placeholder="Write your comment here..."
      value=""
    />
    <span
      className="message"
    >
      Comment cannot be empty
    </span>
  </div>
  <button
    className="full"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`the <CommentForm /> component should render comment form to reply to a comment 1`] = `
.c0 {
  margin: 2rem auto;
}

.c0 p {
  font-size: 1.25rem;
  line-height: 1.75rem;
}

.c0 div {
  margin: 2rem 0;
}

.c0 div input[type='text'],
.c0 div textarea {
  width: 100% !important;
  border-radius: 0.4rem;
  box-sizing: border-box;
  margin: 1rem 0;
  box-shadow: 0 0 20px 0 rgba(0,0,0,0.2),0 5px 5px 0 rgba(0,0,0,0.24);
  padding: 0.5rem;
  display: block;
}

.c0 div input[type='text']:focus-visible,
.c0 div textarea:focus-visible,
.c0 div input[type='text']:focus:not(:focus-visible),
.c0 div textarea:focus:not(:focus-visible) {
  outline: none;
  border: 2px solid;
}

.c0 div input[type='text'] {
  height: 2rem;
}

.c0 div textarea {
  height: 10rem !important;
  resize: none;
}

.c0 div .invalid {
  border: 2px solid !important;
}

.c0 div .message {
  font-size: 0.75rem;
  display: none;
}

.c0 div .show {
  display: block !important;
}

@media all and (min-width:960px) {
  .c0 {
    text-align: left;
  }

  .c0 div input[type='text'],
  .c0 div textarea {
    width: 70% !important;
  }
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <h2>
    Submit a comment
  </h2>
  <p>
    Reply to test
  </p>
  <div>
    <input
      className=""
      id="user"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <span
      className="message"
    >
      Username cannot be empty
    </span>
  </div>
  <div>
    <textarea
      className=""
      id="content"
      name="comment"
      onChange={[Function]}
      placeholder="Write your comment here..."
      value=""
    />
    <span
      className="message"
    >
      Comment cannot be empty
    </span>
  </div>
  <button
    className="full"
    type="submit"
  >
    Submit
  </button>
</form>
`;
